gn3/api/rqtl2.py

```diff
@@ -1,47 +1,47 @@
 """ File contains endpoints for rqlt2"""
-
-import subprocess
-import os
+import shutil
+from pathlib import Path
 from flask import current_app
 from flask import jsonify
 from flask import Blueprint
 from flask import request
-
+from gn3.computations.rqtl2 import compose_rqtl2_cmd
+from gn3.computations.rqtl2 import prepare_files
+from gn3.computations.rqtl2 import validate_required_keys
+from gn3.computations.rqtl2 import write_input_file
+from gn3.computations.rqtl2 import process_qtl2_results
+from gn3.computations.streaming import run_process
 rqtl2 = Blueprint("rqtl2", __name__)
 
 
-@rqtl2.route("/compute", methods=["GET"])
+@rqtl2.route("/compute", methods=["POST"])
 def compute():
     """Endpoint for computing QTL analysis using R/QTL2"""
-    # get the run id to act as file identifier default to output
+    data = request.json
+    required_keys = ["crosstype", "geno_data","pheno_data", "geno_codes"]
+    valid, error = validate_required_keys(required_keys,data)
+    if not valid:
+        return jsonify({"Error" : error}), 400
+    # Provide atleast one  of this data entries.
+    if "physical_map_data" not in data and "geno_map_data" not in data:
+        return jsonify({ "Error":"You need to Provide\
+        Either the Physical map or Geno Map data of markers"}), 400
     run_id = request.args.get("id", "output")
-    output_file = os.path.join(current_app.config.get("TMPDIR"),
-                               f"{run_id}.txt")
-    # this should be computed locally not via files
-    rscript_cmd = (
-        "Rscript ./scripts/rqtl2_wrapper.R "
-        "-i /home/kabui/r_playground/meta_grav.json "
-        "-d /home/kabui/r_playground "
-        "-o /home/kabui/r_playground/rqtl_output.json "
-        "--nperm 100  --threshold 1 --cores 0"
-    )
-    # pylint: disable=consider-using-with
-    process = subprocess.Popen(
-        rscript_cmd, shell=True,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT
-    )
-    for line in iter(process.stdout.readline, b""):
-        # these allow endpoint stream to read the file since
-        # no read and write file same tiem
-        with open(output_file, "a+", encoding="utf-8") as file_handler:
-            file_handler.write(line.decode("utf-8"))
-    process.stdout.close()
-    process.wait()
-    if process.returncode == 0:
-        return jsonify({"msg": "success",
-                        "results": "file_here",
-                        "run_id": run_id})
-    return jsonify({"msg": "fail",
-                    "error": "Process failed",
-                    "run_id": run_id})
+    # prepare necessary files and dir for computation
+    (workspace_dir, input_file,
+     output_file, log_file) = prepare_files(current_app.config.get("TMPDIR"))
+    # write the input file with data required for creating the cross
+    write_input_file(input_file, workspace_dir, data)
+    # TODO fix this
```

Comment:
*nitpick: issue:* Provide more context in this KLUDGE "TODO" to make it easier on someone else looking at this.